### ISSUE TITLE
Segment control for complex devices

### ISSUE THREAD

**Issue by rixls** [wez/govee2mqtt#105]
### Govee Device SKU

H6046, H61E0

### Govee2MQTT Version

2024.01.21-33e8cd7c

## TL;DR:

- Some devices support the 15-segment API when inappropriate. Sometimes the app even promotes it.
- `op.command[]` from MQTT needs to be decoded to reliably work with segmented lights like these. 

---

Like #96, I have segmented devices that misleadingly expose 15 segments through the API.

A pair of TV light bars with 10 total segments ([H6046](https://us.govee.com/products/govee-rgbic-tv-light-bars-for-55-75-inch-tvs)) configured through a single interface:
<img src="https://github.com/wez/govee2mqtt/assets/140927018/73aacb8a-cbcd-495d-8352-d4be7cd9b6d4" width="320">

and a strip light with 20 segments ([H61E0](https://us.govee.com/products/govee-led-strip-light-m1-matter-compatible)) with "segmented" and "advanced" interfaces:

<img src="https://github.com/wez/govee2mqtt/assets/140927018/d599c0c1-b494-4de6-9919-9b2583d36105" width="320">

<img src="https://github.com/wez/govee2mqtt/assets/140927018/453e1c9c-9517-43b3-8570-371ef1d734d4" width="320">

The 15-segment config of the latter seems to be what's exposed through the API. Using the API, the first 10 elements map to the 10 physical segments of the H6046. On the H61E0, all 15 elements are interpolated across the 20 segments.

I isolated the H6046 on a new account and analyzed some traffic. I'm including a few snippets to compare to your own devices. I hope this info is helpful and at least minimally novel.

A request to power on the whole device (both light bars) sends a message like this to MQTT:

```
> GD/<uuid>{"msg":{"accountTopic":"GA/<uuid>","cmd":"turn","cmdVersion":0,"data":{"subVal":[1,1],"val":1},"transaction":"u_0000000000000","type":1}}
< GA/<uuid>{"proType":2,"sku":"H6046","device":"DE:AD:BE:EF:CA:FE:BA:BE","softVersion":"1.00.09","wifiSoftVersion":"2.05.08","wifiHardVersion":"1.02.00","cmd":"status","type":0,"transaction":"u_0000000000000","pactType":1,"pactCode":1,"state":{"onOff":0,"brightness":42,"color":{"r":255,"g":0,"b":0},"colorTemInKelvin":0,"mode":21,"result":1},"op":{"command":["qgUVAAAAAAAAAAAAAAAAAAAAALo=","qqUBZP8AAGT/AABkAAAAZP8AAPE=","qqUCZP8AAGT/AABk/wAAZAAAAPI=","qqUDZP8AAGT/AAAPZGRkAGRkZAM=","qhEAHg8PAAAAAAAAAAAAAAAAAKU=","qhL/ZAAAgAoAAAAAAAAAAAAAAKk=","qiP/AAAAgAAAAIAAAACAAAAAgHY=","qjMAAAAAAAAAAAAAAAAAAAAAAJk="]}}
```

The binary device state returned describes the correct number of bars and segments:

```
aa 05 15 | 00 23 28 23 28 00 00 00 00 00 00 00 00 00 00 00 | ba
aa a5 01 | 64 ff ff ff 64 ff ff ff 64 ff ff ff 64 ff ff ff | 0e
aa a5 02 | 64 ff ff ff 64 ff ff ff 64 ff ff ff 64 ff ff ff | 0d
aa a5 03 | 64 ff ff ff 64 ff ff ff 0f 64 64 64 00 64 64 64 | 03
aa 11 00 | 1e 0f 0f 00 00 00 00 00 00 00 00 00 00 00 00 00 | a5
aa 12 ff | 64 00 00 80 0a 00 00 00 00 00 00 00 00 00 00 00 | a9
aa 23 ff | 00 00 00 80 00 00 00 80 00 00 00 80 00 00 00 80 | 76
aa 33 11 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | 88
```

Each line of the descriptor has an apparent 3-byte header, 16-byte payload, and whatever the last byte is. Above, my two bars are set to a temperature of 0x2328 (9000K) with the ten RGB segments at 0x64 (100%) relative brightness and 0xFFFFFF color. I haven't investigated all the empty bytes, which might be used for DIY mode, DreamView, etc.

This interface seems to be the only feature-complete way to manage these devices.

### Startup Diagnostics

<details>
[2024-01-22T16:26:22 INFO  govee::commands::serve] Light Bars (DE:AD:BE:EF:CA:FE:BA:BE H6046)
[2024-01-22T16:26:22 INFO  govee::commands::serve]   LAN API: ip=192.168.103.118
[2024-01-22T16:26:22 INFO  govee::commands::serve]   Platform API: devices.types.light. supports_rgb=true supports_brightness=true
[2024-01-22T16:26:22 INFO  govee::commands::serve]                 color_temp=Some((2000, 9000)) segment_rgb=Some(0..15)
[2024-01-22T16:26:22 INFO  govee::commands::serve]   Undoc: room=None supports_iot=true ble_only=false
[2024-01-22T16:26:22 INFO  govee::commands::serve]   Quirk { sku: "H6046", icon: "mdi:television-ambient-light", supports_rgb: true, supports_brightness: true, color_temp_range: Some((2000, 9000)), avoid_platform_api: false, ble_only: false, lan_api_capable: true, device_type: Light, platform_temperature_sensor_units: None, platform_humidity_sensor_units: None, iot_api_supported: true, show_as_preset_buttons: None }
[2024-01-22T16:26:22 INFO  govee::commands::serve] 
[2024-01-22T16:26:22 INFO  govee::commands::serve] Light Strip (DE:AD:BE:EF:CA:FE:BA:BE H61E0)
[2024-01-22T16:26:22 INFO  govee::commands::serve]   LAN API: ip=192.168.103.117
[2024-01-22T16:26:22 INFO  govee::commands::serve]   Platform API: devices.types.light. supports_rgb=true supports_brightness=true
[2024-01-22T16:26:22 INFO  govee::commands::serve]                 color_temp=Some((2000, 9000)) segment_rgb=Some(0..15)
[2024-01-22T16:26:22 INFO  govee::commands::serve]   Undoc: room=None supports_iot=true ble_only=false
[2024-01-22T16:26:22 INFO  govee::commands::serve]   Quirk { sku: "H61E0", icon: "mdi:light-bulb", supports_rgb: true, supports_brightness: true, color_temp_range: Some((2000, 9000)), avoid_platform_api: false, ble_only: false, lan_api_capable: true, device_type: Light, platform_temperature_sensor_units: None, platform_humidity_sensor_units: None, iot_api_supported: true, show_as_preset_buttons: None }
</details>

### Additional Logs

<details><summary>H6046 metadata</summary>

```
{
  "meta": {
    "capabilities": [
      {
        "alarmType": null,
        "eventState": null,
        "instance": "powerSwitch",
        "parameters": {
          "dataType": "ENUM",
          "options": [
            {
              "name": "on",
              "value": 1
            },
            {
              "name": "off",
              "value": 0
            }
          ]
        },
        "type": "devices.capabilities.on_off"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "gradientToggle",
        "parameters": {
          "dataType": "ENUM",
          "options": [
            {
              "name": "on",
              "value": 1
            },
            {
              "name": "off",
              "value": 0
            }
          ]
        },
        "type": "devices.capabilities.toggle"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "brightness",
        "parameters": {
          "dataType": "INTEGER",
          "range": {
            "max": 100,
            "min": 1,
            "precision": 1
          },
          "unit": "unit.percent"
        },
        "type": "devices.capabilities.range"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "segmentedBrightness",
        "parameters": {
          "dataType": "STRUCT",
          "fields": [
            {
              "dataType": "Array",
              "defaultValue": null,
              "elementRange": {
                "max": 14,
                "min": 0
              },
              "elementType": "INTEGER",
              "fieldName": "segment",
              "options": [],
              "required": true,
              "size": {
                "max": 15,
                "min": 1
              }
            },
            {
              "dataType": "INTEGER",
              "defaultValue": null,
              "fieldName": "brightness",
              "range": {
                "max": 100,
                "min": 0,
                "precision": 1
              },
              "required": true,
              "unit": null
            }
          ]
        },
        "type": "devices.capabilities.segment_color_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "segmentedColorRgb",
        "parameters": {
          "dataType": "STRUCT",
          "fields": [
            {
              "dataType": "Array",
              "defaultValue": null,
              "elementRange": {
                "max": 14,
                "min": 0
              },
              "elementType": "INTEGER",
              "fieldName": "segment",
              "options": [],
              "required": true,
              "size": {
                "max": 15,
                "min": 1
              }
            },
            {
              "dataType": "INTEGER",
              "defaultValue": null,
              "fieldName": "rgb",
              "range": {
                "max": 16777215,
                "min": 0,
                "precision": 1
              },
              "required": true,
              "unit": null
            }
          ]
        },
        "type": "devices.capabilities.segment_color_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "colorRgb",
        "parameters": {
          "dataType": "INTEGER",
          "range": {
            "max": 16777215,
            "min": 0,
            "precision": 1
          },
          "unit": null
        },
        "type": "devices.capabilities.color_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "colorTemperatureK",
        "parameters": {
          "dataType": "INTEGER",
          "range": {
            "max": 9000,
            "min": 2000,
            "precision": 1
          },
          "unit": null
        },
        "type": "devices.capabilities.color_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "lightScene",
        "parameters": {
          "dataType": "ENUM",
          "options": []
        },
        "type": "devices.capabilities.dynamic_scene"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "musicMode",
        "parameters": {
          "dataType": "STRUCT",
          "fields": [
            {
              "dataType": "ENUM",
              "defaultValue": null,
              "fieldName": "musicMode",
              "options": [
                {
                  "name": "Vivid",
                  "value": 1
                },
                {
                  "name": "Rhythm",
                  "value": 2
                },
                {
                  "name": "Strike",
                  "value": 3
                },
                {
                  "name": "Vibrate",
                  "value": 4
                },
                {
                  "name": "Beat",
                  "value": 5
                },
                {
                  "name": "Torch",
                  "value": 6
                },
                {
                  "name": "RainbowCircle",
                  "value": 7
                },
                {
                  "name": "Shiny",
                  "value": 8
                }
              ],
              "required": true
            },
            {
              "dataType": "INTEGER",
              "defaultValue": null,
              "fieldName": "sensitivity",
              "range": {
                "max": 100,
                "min": 0,
                "precision": 1
              },
              "required": true,
              "unit": "unit.percent"
            },
            {
              "dataType": "ENUM",
              "defaultValue": null,
              "fieldName": "autoColor",
              "options": [
                {
                  "name": "on",
                  "value": 1
                },
                {
                  "name": "off",
                  "value": 0
                }
              ],
              "required": false
            },
            {
              "dataType": "INTEGER",
              "defaultValue": null,
              "fieldName": "rgb",
              "range": {
                "max": 16777215,
                "min": 0,
                "precision": 1
              },
              "required": false,
              "unit": null
            }
          ]
        },
        "type": "devices.capabilities.music_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "diyScene",
        "parameters": {
          "dataType": "ENUM",
          "options": []
        },
        "type": "devices.capabilities.dynamic_scene"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "snapshot",
        "parameters": {
          "dataType": "ENUM",
          "options": []
        },
        "type": "devices.capabilities.dynamic_scene"
      }
    ],
    "device": "_",
    "deviceName": "_",
    "sku": "H6046",
    "type": "devices.types.light"
  },
  "state": null
}
```

</details>

<details><summary>H610E metadata</summary>

```
{
  "meta": {
    "capabilities": [
      {
        "alarmType": null,
        "eventState": null,
        "instance": "powerSwitch",
        "parameters": {
          "dataType": "ENUM",
          "options": [
            {
              "name": "on",
              "value": 1
            },
            {
              "name": "off",
              "value": 0
            }
          ]
        },
        "type": "devices.capabilities.on_off"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "gradientToggle",
        "parameters": {
          "dataType": "ENUM",
          "options": [
            {
              "name": "on",
              "value": 1
            },
            {
              "name": "off",
              "value": 0
            }
          ]
        },
        "type": "devices.capabilities.toggle"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "brightness",
        "parameters": {
          "dataType": "INTEGER",
          "range": {
            "max": 100,
            "min": 1,
            "precision": 1
          },
          "unit": "unit.percent"
        },
        "type": "devices.capabilities.range"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "segmentedBrightness",
        "parameters": {
          "dataType": "STRUCT",
          "fields": [
            {
              "dataType": "Array",
              "defaultValue": null,
              "elementRange": {
                "max": 14,
                "min": 0
              },
              "elementType": "INTEGER",
              "fieldName": "segment",
              "options": [],
              "required": true,
              "size": {
                "max": 15,
                "min": 1
              }
            },
            {
              "dataType": "INTEGER",
              "defaultValue": null,
              "fieldName": "brightness",
              "range": {
                "max": 100,
                "min": 0,
                "precision": 1
              },
              "required": true,
              "unit": null
            }
          ]
        },
        "type": "devices.capabilities.segment_color_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "segmentedColorRgb",
        "parameters": {
          "dataType": "STRUCT",
          "fields": [
            {
              "dataType": "Array",
              "defaultValue": null,
              "elementRange": {
                "max": 14,
                "min": 0
              },
              "elementType": "INTEGER",
              "fieldName": "segment",
              "options": [],
              "required": true,
              "size": {
                "max": 15,
                "min": 1
              }
            },
            {
              "dataType": "INTEGER",
              "defaultValue": null,
              "fieldName": "rgb",
              "range": {
                "max": 16777215,
                "min": 0,
                "precision": 1
              },
              "required": true,
              "unit": null
            }
          ]
        },
        "type": "devices.capabilities.segment_color_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "colorRgb",
        "parameters": {
          "dataType": "INTEGER",
          "range": {
            "max": 16777215,
            "min": 0,
            "precision": 1
          },
          "unit": null
        },
        "type": "devices.capabilities.color_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "colorTemperatureK",
        "parameters": {
          "dataType": "INTEGER",
          "range": {
            "max": 9000,
            "min": 2000,
            "precision": 1
          },
          "unit": null
        },
        "type": "devices.capabilities.color_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "lightScene",
        "parameters": {
          "dataType": "ENUM",
          "options": []
        },
        "type": "devices.capabilities.dynamic_scene"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "musicMode",
        "parameters": {
          "dataType": "STRUCT",
          "fields": [
            {
              "dataType": "ENUM",
              "defaultValue": null,
              "fieldName": "musicMode",
              "options": [
                {
                  "name": "Energic",
                  "value": 1
                },
                {
                  "name": "Rhythm",
                  "value": 2
                },
                {
                  "name": "Spectrum",
                  "value": 3
                },
                {
                  "name": "Rolling",
                  "value": 4
                },
                {
                  "name": "Separation",
                  "value": 5
                },
                {
                  "name": "Hopping",
                  "value": 6
                },
                {
                  "name": "PianoKeys",
                  "value": 7
                },
                {
                  "name": "Fountain",
                  "value": 8
                },
                {
                  "name": "DayAndNight",
                  "value": 9
                },
                {
                  "name": "Sprouting",
                  "value": 10
                },
                {
                  "name": "Shiny",
                  "value": 11
                }
              ],
              "required": true
            },
            {
              "dataType": "INTEGER",
              "defaultValue": null,
              "fieldName": "sensitivity",
              "range": {
                "max": 100,
                "min": 0,
                "precision": 1
              },
              "required": true,
              "unit": "unit.percent"
            },
            {
              "dataType": "ENUM",
              "defaultValue": null,
              "fieldName": "autoColor",
              "options": [
                {
                  "name": "on",
                  "value": 1
                },
                {
                  "name": "off",
                  "value": 0
                }
              ],
              "required": false
            },
            {
              "dataType": "INTEGER",
              "defaultValue": null,
              "fieldName": "rgb",
              "range": {
                "max": 16777215,
                "min": 0,
                "precision": 1
              },
              "required": false,
              "unit": null
            }
          ]
        },
        "type": "devices.capabilities.music_setting"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "diyScene",
        "parameters": {
          "dataType": "ENUM",
          "options": []
        },
        "type": "devices.capabilities.dynamic_scene"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "snapshot",
        "parameters": {
          "dataType": "ENUM",
          "options": [
            {
              "name": "Christmas 1",
              "value": 751271
            },
            {
              "name": "work light",
              "value": 751278
            },
            {
              "name": "Neutral",
              "value": 764430
            },
            {
              "name": "Christmas",
              "value": 868110
            }
          ]
        },
        "type": "devices.capabilities.dynamic_scene"
      },
      {
        "alarmType": null,
        "eventState": null,
        "instance": "dreamViewToggle",
        "parameters": {
          "dataType": "ENUM",
          "options": [
            {
              "name": "on",
              "value": 1
            },
            {
              "name": "off",
              "value": 0
            }
          ]
        },
        "type": "devices.capabilities.toggle"
      }
    ],
    "device": "_",
    "deviceName": "_",
    "sku": "H61E0",
    "type": "devices.types.light"
  },
  "state": null
}
```

</details>

### Anything else?

<details><summary>Some extra H6046 captures</summary>

```
# binary segment change
> GD/<uuid>{"msg":{"accountTopic":"GA/<uuid>","cmd":"ptReal","cmdVersion":0,"data":{"command":["<base64 blob>\u003d"]},"transaction":"u_0000000000000","type":1}}
< GA/<uuid>{"proType":2,"sku":"H6046","device":"DE:AD:BE:EF:CA:FE:BA:BE","softVersion":"1.00.09","wifiSoftVersion":"2.05.08","wifiHardVersion":"1.02.00","cmd":"ptReal","type":1,"transaction":"u_0000000000000","pactType":1,"pactCode":1,"state":{"result":1},"op":{"command":["MwUAAAAAAAAAAAAAAAAAAAAAADY="]}}
```

```
# text segment change (whole device)
> GD/<uuid>{"msg":{"accountTopic":"GA/<uuid>","cmd":"colorwc","cmdVersion":0,"data":{"color":{"b":255,"g":255,"r":255},"colorTemInKelvin":0},"transaction":"u_0000000000000","type":1}}
< GA/<uuid>{"proType":2,"sku":"H6046","device":"DE:AD:BE:EF:CA:FE:BA:BE","softVersion":"1.00.09","wifiSoftVersion":"2.05.08","wifiHardVersion":"1.02.00","cmd":"colorwc","type":0,"transaction":"u_0000000000000","pactType":1,"pactCode":1,"state":{"color":{"r":255,"g":0,"b":0},"colorTemInKelvin":0,"result":1}}
```

Byte|Thing
-|-
0|The first byte, 0x33
1|Command domain (led=>0x05, power=>0x33)
2-18|Payload
19|The twentieth byte

Byte|LED payload
-|-
2|Mode (color=>0x15, scene=>0x04)
3|Command (RGB=>0x01, brightness=>0x02)
4-19|Config

Byte|Color config
-|-
4-6|RGB
7-11|Temp
12-??|LE segment bitmask

```
# set seg 1-10 to preset red (0xff0000)
33 05 15 01 | ff 00 00 | 00 00 00 00 00 | 01 00 00 00 00 00 00 | dc

# set seg 1 to color wheel pink (0xff00fc)
33 05 15 01 | ff 00 fc | 00 00 00 00 00 | 01 00 00 00 00 00 00 | 20

# set segs 1-9 to preset red (0xff0000) simultaneously
33 05 15 01 | ff 00 00 | 00 00 00 00 00 | ff 01 00 00 00 00 00 | 23

# set seg 1 (whole left bar) color to warmest white (2100K)
33 05 15 01 | ff ff ff | 08 34 ff 92 1d | 01 00 00 00 00 00 00 | 90

# set seg 6 (whole right bar) color to coolest white (9000K)
33 05 15 01 | ff ff ff | 22 60 db e2 ff | 20 00 00 00 00 00 00 | 79
```

Byte|Brightness config
-|-
4|Brightness
5-??|LE segment bitmask

```
# set seg 1 rel brightness to 84% (0x54)
33 05 15 02 | 54 | 01 00 00 00 00 00 00 00 00 00 00 00 00 00 | 74

# set seg 10 rel brightness to 54% (0x36)
33 05 15 02 | 36 | 00 02 00 00 00 00 00 00 00 00 00 00 00 00 | 15
```

Byte|Power payload
-|-
2|On/off mask

```
# power on left bar only
33 33 | 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | 01

# power on right bar only
33 33 | 10 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | 10

# power on both bars
33 33 | 11 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | 11

# power off both bars
33 33 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | 00
```

Scene mode:

```
# switch to scene mode
33 05 04 | 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | 2a

# set scene to "snow flake"
33 05 04 | 0f 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | 3d
```
</details>

**Comment by wez** [comment]
Thanks for the info!
Can you share some details on how you captured those packets?

**Comment by rixls** [comment]
1.  Load Govee Home in Bluestacks 5 (Android 11, rooted)
2.  Set up Frida to [proxy 443, 8883 and disable certificate pinning](https://github.com/httptoolkit/frida-interception-and-unpinning) 
\
Nope runs on its own port so I tweaked the proxy script to pass 443 to Burp and 8883 to Nope. I also needed an additional unpinning script:
```Java
Java.perform(function () {
    var TrustManagerImpl = Java.use('com.android.org.conscrypt.TrustManagerImpl');
    var ArrayList = Java.use("java.util.ArrayList");
    TrustManagerImpl.verifyChain.implementation = function(untrustedChain, trustAnchorChain,
        host, clientAuth, ocspData, tlsSctData) {
        console.log("[+] Bypassing TrustManagerImpl->verifyChain()");
        return untrustedChain;
    }
    TrustManagerImpl.checkTrustedRecursive.implementation = function(certs, host, clientAuth, untrustedChain,
        trustAnchorChain, used) {
        console.log("[+] Bypassing TrustManagerImpl->checkTrustedRecursive()");
        return ArrayList.$new();
    };
});
```
3.  Inspect the traffic with Burp and Nope ([lightly modified to load the MQTT client cert](https://github.com/summitt/Nope-Proxy/blob/master/NonHTTPProxy/src/josh/nonHttp/GenericMiTMServer.java#L285).)
\
The quickest, dirtiest mod:
``` Diff
-            			sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
+						KeyStore keyStore = KeyStore.getInstance("PKCS12");
+						String password = "[P12 PASSWORD]";
+						try (InputStream in = new FileInputStream("D:/path/to/client.p12")) {
+							keyStore.load(in, password.toCharArray());
+						}
+						KeyManagerFactory kmf = KeyManagerFactory.getInstance("X509");
+						kmf.init(keyStore, password.toCharArray());
+						sslContext.init(kmf.getKeyManagers(), trustAllCerts, new java.security.SecureRandom());
```

There shouldn't be any more to it than that :)
